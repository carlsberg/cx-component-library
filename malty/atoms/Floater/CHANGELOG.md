# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- additional stories ([#291](https://github.com/CarlsbergGBS/cx-component-library/issues/291)) ([438af3b](https://github.com/CarlsbergGBS/cx-component-library/commit/438af3b969e0f5b4b908436ee58b93c4b9ddc97c))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-229:** button ([#165](https://github.com/CarlsbergGBS/cx-component-library/issues/165)) ([1950717](https://github.com/CarlsbergGBS/cx-component-library/commit/19507172b731f341ecf32c13eac07e465c6aa659))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-338:** themeable floater ([#293](https://github.com/CarlsbergGBS/cx-component-library/issues/293)) ([a855bd3](https://github.com/CarlsbergGBS/cx-component-library/commit/a855bd3f44472675172c950cbf268685c1a8391c))
- **DSM-338:** themeable floater ([#296](https://github.com/CarlsbergGBS/cx-component-library/issues/296)) ([13bc5d0](https://github.com/CarlsbergGBS/cx-component-library/commit/13bc5d00ded44cc202625ccfceae72c549f354c1))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **floater:** [DSM-710] Add tests ([#704](https://github.com/CarlsbergGBS/cx-component-library/issues/704)) ([e23364b](https://github.com/CarlsbergGBS/cx-component-library/commit/e23364bf8632e000b2563634af450fed2bb83687))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-775] storybook v7 upgrade ([#782](https://github.com/CarlsbergGBS/cx-component-library/issues/782)) ([23eee21](https://github.com/CarlsbergGBS/cx-component-library/commit/23eee211673a2351970d70e9f4533f135840540b))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
