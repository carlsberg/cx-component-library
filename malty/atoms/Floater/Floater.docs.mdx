---
labels: ['react', 'typescript', 'floater']
description: 'A Floater component.'
---
