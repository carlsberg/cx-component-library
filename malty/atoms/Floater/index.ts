export { Floater } from './Floater';
export { FloaterColor, FloaterIconPosition } from './Floater.types';
export type { FloaterProps } from './Floater.types';
