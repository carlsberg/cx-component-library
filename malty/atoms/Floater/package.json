{
  "name": "@carlsberg/malty.atoms.floater",
  "version": "0.1.5",
  "main": "./lib/cjs/index.js",
  "module": "./lib/esm/index.js",
  "types": "./lib/esm/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/carlsberg/malty.git"
  },
  "publishConfig": {
    "@carlsberg:registry": "https://npm.pkg.github.com",
    "access": "public"
  },
  "scripts": {
    "build": "yarn build:esm && yarn build:cjs",
    "build:esm": "tsc",
    "build:cjs": "tsc --module commonjs --outDir lib/cjs",
    "test": "jest"
  },
  "dependencies": {
    "@carlsberg/malty.atoms.base-icon": "^0.0.4",
    "@carlsberg/malty.theme.malty-theme-provider": "^0.0.45",
    "@carlsberg/malty.utils.space": "^0.0.4",
    "core-js": "^3.0.0",
    "styled-components": "^5.3.6"
  },
  "devDependencies": {
    "@babel/runtime": "7.12.18",
    "@carlsberg/malty.icons.arrow-small-up": "^0.1.2",
    "@carlsberg/malty.utils.all-icons": "^0.0.4",
    "@carlsberg/malty.utils.test": "^0.0.20",
    "@storybook/react": "^7.4.1",
    "@testing-library/react": "^12.1.5",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^26.0.0",
    "@types/node": "12.20.4",
    "@types/react": "^17.0.8",
    "@types/react-dom": "^17.0.5",
    "@types/styled-components": "^5.1.26",
    "@types/testing-library__jest-dom": "5.9.5"
  },
  "peerDependencies": {
    "react": "^16.8.0 || ^17.0.0",
    "react-dom": "^16.8.0 || ^17.0.0"
  }
}
