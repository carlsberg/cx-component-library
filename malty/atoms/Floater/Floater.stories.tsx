import { ArrowSmallUp } from '@carlsberg/malty.icons.arrow-small-up';
import { allIconsStoryOptions } from '@carlsberg/malty.utils.all-icons';
import { generateStorybookSpacing } from '@carlsberg/malty.utils.space';
import { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import { Floater } from './Floater';
import { FloaterColor, FloaterIconPosition, FloaterProps } from './Floater.types';

const StyledWrapper = styled.div`
  height: 50px;
`;

const meta: Meta<FloaterProps> = {
  title: 'Forms/Floater',
  component: Floater,
  parameters: {
    importObject: 'Floater',
    importPath: '@carlsberg/malty.atoms.floater'
  },
  render: (args) => (
    <StyledWrapper>
      <Floater {...args} />
    </StyledWrapper>
  ),
  argTypes: {
    text: {
      control: 'text',
      description: 'Button label, can be'
    },
    onClick: {
      description: 'This is a function that will run on click. It is not a required property',
      table: {
        category: 'Events'
      }
    },
    scroll: {
      description: 'Scroll position where will floater show',
      table: {
        defaultValue: {
          summary: 0
        }
      },
      control: 'number'
    },
    negative: {
      control: 'boolean',
      description: 'Should this be a white button?',
      table: {
        category: 'Styling'
      }
    },
    icon: {
      description: 'The icon component to be displayed',
      options: allIconsStoryOptions,
      control: 'select',
      table: {
        category: 'Icon'
      }
    },
    iconPos: {
      description: 'When icon present, position will be',
      control: 'select',
      mapping: FloaterIconPosition,
      options: Object.keys(FloaterIconPosition),
      table: {
        category: 'Icon',
        defaultValue: {
          summary: FloaterIconPosition.Right
        }
      }
    },
    color: {
      description: 'Floater color. Options are',
      control: 'select',
      mapping: FloaterColor,
      options: Object.keys(FloaterColor),
      table: {
        category: 'Styling',
        defaultValue: {
          summary: FloaterColor.DigitalBlack
        }
      }
    },
    dataTestId: {
      control: 'text',
      description: 'Floater data-testid'
    },
    ...generateStorybookSpacing()
  }
};

type Story = StoryObj<FloaterProps>;

export const Base: Story = {
  args: {
    text: '',
    negative: false,
    icon: <ArrowSmallUp />,
    iconPos: FloaterIconPosition.Right
  }
};

export const Text: Story = {
  args: {
    ...Base.args,
    text: 'Floater'
  }
};

export default meta;
