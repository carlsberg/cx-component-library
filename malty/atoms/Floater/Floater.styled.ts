import styled, { css } from 'styled-components';
import { FloaterIconPosition } from './Floater.types';

export const StyledFloaterButton = styled.button<{
  hasText: boolean;
  hasIcon: boolean;
  isNegative?: boolean;
  fullWidth?: boolean;

  iconPos: FloaterIconPosition;
  showButton: boolean;
}>`
  background-color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default.white.value : theme.colors.theme.themePrimary.value};
  color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.default.white.value};
  font-family: inherit;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: ${({ theme }) => `0 ${theme.sizes.s.value}`};
  height: ${({ theme }) => `${theme.sizes['2xl'].value}`};
  font-size: ${({ theme }) => theme.typography.desktop.text['medium-small_bold']['font-size'].value};
  font-weight: ${({ theme }) => theme.typography.desktop.text['medium-small_bold']['font-weight'].value};
  z-index: 100;
  right: ${({ theme }) => theme.sizes.m.value};
  position: absolute;
  bottom: ${({ theme, showButton }) => (showButton ? theme.sizes.m.value : '-36px')};
  transition: 0.25s ease-in-out;
  transition-property: background-color, color;
  cursor: pointer;
  border: none;
  border-radius: ${({ theme }) => `${theme.sizes.m.value}`};
  gap: ${({ theme }) => theme.sizes.s.value};
  width: auto;
  flex-direction: ${({ iconPos }) =>
    FloaterIconPosition[iconPos] === FloaterIconPosition.Right ? 'row' : 'row-reverse'};

  @media screen and (max-width: ${({ theme }) => theme.layout.small['device-max-width']?.value}) {
    bottom: ${({ theme, showButton }) => (showButton ? theme.sizes.s.value : '-36px')};
    right: ${({ theme }) => theme.sizes.s.value};
  }
  &:hover {
    opacity: 0.95;
  }
  &:active {
    opacity: 0.9;
  }

  .text-container {
    display: flex;
    align-items: center;
    gap: ${({ theme }) => theme.sizes.s.value};
    opacity: 1;
  }

  ${({ hasText, hasIcon }) =>
    !hasText &&
    hasIcon &&
    css`
      padding: 0;
      justify-content: center;

      width: ${({ theme }) => `${theme.sizes['2xl'].value}`};
    `};
`;
