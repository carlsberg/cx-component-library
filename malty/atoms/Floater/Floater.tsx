import { CloneIcon, IconColor, IconSize } from '@carlsberggbs/malty.atoms.base-icon';
import { globalTheme as defaultTheme } from '@carlsberggbs/malty.theme.malty-theme-provider';
import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { FloaterIconPosition, FloaterProps } from '.';
import { StyledFloaterButton } from './Floater.styled';
import { FloaterColor } from './Floater.types';

export const Floater = ({
  text,
  onClick,
  icon,
  iconPos = FloaterIconPosition.Right,
  scroll,
  negative,
  tabIndex = -1,
  dataTestId,
  children,
  color = FloaterColor.DigitalBlack,
  ...props
}: FloaterProps) => {
  let iconColor;
  const theme = useContext(ThemeContext) || defaultTheme;
  const [showButton, setShowButton] = useState(true);

  switch (color) {
    case FloaterColor.ThemePrimary:
      iconColor = IconColor.White;
      if (negative) {
        iconColor = IconColor.Primary;
      }
      break;
    case FloaterColor.ThemeSecondary:
      iconColor = IconColor.White;
      if (negative) {
        iconColor = IconColor.Secondary;
      }
      break;
    case FloaterColor.ThemeTertiary:
      iconColor = IconColor.White;
      if (negative) {
        iconColor = IconColor.Tertiary;
      }
      break;

    default:
      iconColor = negative ? IconColor.DigitalBlack : IconColor.White;
      break;
  }

  const handleScroll = () => {
    const st = window.pageYOffset || document.documentElement.scrollTop;
    if (!!scroll && st > scroll) {
      setShowButton(true);
    } else {
      setShowButton(false);
    }
  };

  useEffect(() => {
    if (scroll) {
      setShowButton(false);
      document.addEventListener('scroll', handleScroll);
    }
    return () => {
      document.removeEventListener('scroll', handleScroll);
    };
  }, [scroll]);

  const renderedIcon = <CloneIcon icon={icon} color={iconColor} size={IconSize.Medium} />;

  return (
    <StyledFloaterButton
      color={color}
      data-testid={dataTestId}
      type="button"
      hasText={!!text}
      hasIcon={!!icon}
      onClick={onClick ?? (() => window.scrollTo({ top: 0, behavior: 'smooth' }))}
      isNegative={negative}
      iconPos={iconPos}
      theme={theme}
      showButton={showButton}
      tabIndex={tabIndex}
      {...props}
    >
      <div className="text-container">
        {iconPos === FloaterIconPosition.Left && renderedIcon}
        {text || children}
        {iconPos === FloaterIconPosition.Right && renderedIcon}
      </div>
    </StyledFloaterButton>
  );
};
