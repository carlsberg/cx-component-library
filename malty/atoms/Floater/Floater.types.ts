import { IconName } from '@carlsberggroup/malty.atoms.icon';
import React from 'react';

export interface FloaterProps {
  text?: string | number;
  icon?: IconName;
  iconPos?: FloaterIconPosition;
  negative?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  scroll?: number;
  tabIndex?: number;
  dataTestId?: string;
  children?: string | JSX.Element;
  color?: FloaterColor;
}

export enum FloaterIconPosition {
  Right = 'Right',
  Left = 'Left',
}

export enum FloaterColor {
  DigitalBlack = 'digital-black',
  ThemePrimary = 'themePrimary',
  ThemeSecondary = 'themeSecondary',
  ThemeTertiary = 'themeTertiary',
}
