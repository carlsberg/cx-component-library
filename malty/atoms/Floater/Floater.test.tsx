import { fireEvent, jsonRenderer, render, screen } from '@carlsberggroup/malty.utils.test';
import React from 'react';
import { Floater } from './Floater';

const defaultText = 'Submit';
const newText = 'Go';

describe('floater', () => {
  it('matches snapshot', () => {
    const onClick = jest.fn();
    const view = jsonRenderer(<Floater text={defaultText} onClick={onClick} />);
    expect(view).toMatchSnapshot();
  });

  it('renders with correct text', () => {
    const { rerender } = render(<Floater text={defaultText} />);
    expect(screen.getByText(defaultText)).not.toBeNull();

    rerender(<Floater text={newText} />);
    expect(screen.getByText(newText)).not.toBeNull();
  });

  it('renders with correct text via child', () => {
    render(<Floater>{defaultText}</Floater>);
    expect(screen.getByText(defaultText)).not.toBeNull();
  });

  it('calls function on click', () => {
    const onClick = jest.fn();
    render(<Floater text={defaultText} onClick={onClick} />);
    fireEvent.click(screen.getByText(defaultText));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('does not call function on click when disabled', () => {
    const onClick = jest.fn();
    render(<Floater text={defaultText} onClick={onClick} />);
    fireEvent.click(screen.getByText(defaultText));
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});
