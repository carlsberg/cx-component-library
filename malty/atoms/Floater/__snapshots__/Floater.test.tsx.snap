// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`floater matches snapshot 1`] = `
<span
  className="sc-bczRLJ hROJNc"
>
  <button
    className="sc-dkzDqf eIamyt"
    onClick={[MockFunction]}
    tabIndex={-1}
    type="button"
  >
    <div
      className="text-container"
    >
      Submit
    </div>
  </button>
</span>
`;
